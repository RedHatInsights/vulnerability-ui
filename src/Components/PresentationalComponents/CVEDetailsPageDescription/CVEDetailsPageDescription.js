import { Stack, StackItem, Text, TextContent, TextVariants } from '@patternfly/react-core';
import propTypes from 'prop-types';
import React from 'react';
import { injectIntl } from 'react-intl';
import messages from '../../../Messages';

const CVEDetailsPageDescription = ({ cveAttributes, intl }) => {
    const {
        public_date: publishDate,
        description,
        rh_link: link
    } = cveAttributes;

    return (
        <Stack hasGutter>
            <StackItem className="pf-u-mt-sm">
                <div>
                    {intl.formatMessage(messages.publishDate)}: {publishDate}
                </div>
            </StackItem>
            <StackItem>
                <TextContent>
                    <Text component={TextVariants.p}>{description}</Text>
                </TextContent>
            </StackItem>
            <StackItem className="pf-u-mt-sm pf-u-mb-md">
                {link}
            </StackItem>
        </Stack>

    );
};

CVEDetailsPageDescription.propTypes = {
    cveAttributes: propTypes.any,
    intl: propTypes.any
};

export default injectIntl(CVEDetailsPageDescription);
