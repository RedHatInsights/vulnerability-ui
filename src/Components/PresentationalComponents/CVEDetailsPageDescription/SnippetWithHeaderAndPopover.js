import { Popover, Stack, StackItem } from '@patternfly/react-core';
import propTypes from 'prop-types';
import React from 'react';
import Label from '../Snippets/Label';

const SnippetWithHeaderAndPopover = props => {
    const { title, value, content } = props;

    return (
        <Popover id="popover" bodyContent={content} headerContent={''} position="bottom"
            appendTo={document.querySelector('.vulnerability')}>
            <Stack className="popover-content">
                <StackItem>
                    <Label isLarge>{title}</Label>
                </StackItem>
                <StackItem>
                    <a>{value}</a>
                </StackItem>
            </Stack>
        </Popover>
    );
};

SnippetWithHeaderAndPopover.propTypes = {
    title: propTypes.any,
    value: propTypes.any,
    content: propTypes.any
};
export default SnippetWithHeaderAndPopover;
