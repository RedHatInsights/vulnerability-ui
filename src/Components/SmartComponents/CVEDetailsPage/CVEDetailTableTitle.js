import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import { Skeleton, Text, TextContent, TextVariants } from '@patternfly/react-core';
import { getAffectedSystemsByCVE } from '../../../Helpers/APIHelper';
import messages from '../../../Messages';

const CVEDetailTableTitle = ({ cveName }) => {
    const intl = useIntl();
    const [isLoading, setLoading] = useState(true);
    const [edgeDeviceExist, setEdgeDeviceExist] = useState(false);
    const [totalSystems, setTotalSystems] = useState(0);
    useEffect(() => {
        try {
            getAffectedSystemsByCVE({ id: cveName, host_type: 'edge', limit: 1 }).then(edgeDevice => {
                setEdgeDeviceExist(edgeDevice?.meta.total_items > 0);
                getAffectedSystemsByCVE({ id: cveName }).then(result => {

                    setTotalSystems(result?.meta?.total_items || 0);
                    setLoading(false);
                });
            });
        } catch {
            setEdgeDeviceExist(false);
            setTotalSystems(0);
            setLoading(false);
        }
    }, []);

    return isLoading
        ?
        <Skeleton width="25%" screenreaderText="Loading contents" />
        : (<TextContent aria-label="Affected systems table title">
            <Text component={TextVariants.h2} id="systems-exposed-table-header">
                {edgeDeviceExist
                    ? intl.formatMessage(messages.totalSystemsWithEdge, { count: totalSystems })
                    : intl.formatMessage(messages.affectsSystems)}
            </Text>
        </TextContent>);
};

CVEDetailTableTitle.propTypes = {
    cveName: PropTypes.string.isRequired
};
export default CVEDetailTableTitle;
