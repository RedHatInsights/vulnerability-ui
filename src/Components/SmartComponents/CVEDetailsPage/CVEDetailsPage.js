import { Main } from '@redhat-cloud-services/frontend-components/Main';
import propTypes from 'prop-types';
import React, { useMemo, useState, useEffect } from 'react';
import { createCveDetailsPage } from '../../../Helpers/CVEHelper';
import { useDispatch, useSelector } from 'react-redux';
import {
    fetchCveDetails,
    changeExposedSystemsParameters,
    fetchAffectedSystemsByCVE,
    clearCveStore
} from '../../../Store/Actions/Actions';
import { withRouter } from 'react-router-dom';
import { PATHS } from '../../../Helpers/constants';
import CVEDetailsPageSummary from '../../PresentationalComponents/CVEDetailsPageSummary/CVEDetailsPageSummary';
import { kebabItemEditBusinessRisk, kebabItemEditStatus } from '../../PresentationalComponents/Kebab/KebabItems';
import BusinessRiskModal from '../Modals/BusinessRiskModal';
import CveStatusModal from '../Modals/CveStatusModal';
import SystemsExposedTable from '../SystemsExposedTable/SystemsExposedTable';
import Header from '../../PresentationalComponents/Header/Header';
import GroupedCVELabels from '../../PresentationalComponents/Snippets/GroupedCVELabels';
import ErrorHandler from '../../PresentationalComponents/ErrorHandler/ErrorHandler';

export const CVEPageContext = React.createContext({ isLoading: true });

const CVEDetailsPage = ({ match }) => {
    const dispatch = useDispatch();
    const [cveName] = useState(match.params.cve);
    const [BusinessModal, setBusinessRisk] = useState(() => () => null);
    const [StatusModal, setStatusModal] = useState(() => () => null);

    const details = useSelector(
        ({ CVEDetailsPageStore }) => CVEDetailsPageStore.cveDetails
    );
    const cveDetails = useMemo(() => createCveDetailsPage(details), [details]);

    const exposedSystems = useSelector(
        ({ CVEDetailsPageStore }) => CVEDetailsPageStore.affectedSystemsByCVE
    );

    useEffect(() => {
        dispatch(fetchCveDetails(cveName));
    }, [dispatch, cveName]);

    const showBusinessRiskModal = cves => {
        setBusinessRisk(
            () => () =>
                <BusinessRiskModal open cves={cves} updateRef={() => dispatch(fetchCveDetails(cveName))} />
        );
    };

    const showStatusModal = cves => {
        setStatusModal(
            () => () =>
                <CveStatusModal
                    cves={cves}
                    updateRef={() => {
                        dispatch(fetchCveDetails(cveName));
                        dispatch(fetchAffectedSystemsByCVE(cveName));
                    }}
                />
        );
    };

    useEffect(() => {
        return () => {
            dispatch(clearCveStore());
        };
    }, [dispatch]);

    const { error, data } = cveDetails;

    const cveStatusDetails = {
        id: cveName,
        status_id: cveDetails.data.status_id,
        justification: cveDetails.data.status_justification,
        exposed_systems_count: exposedSystems?.payload?.meta?.total_items || 0
    };
    const kebabItems = [
        kebabItemEditBusinessRisk(showBusinessRiskModal, [
            {
                business_risk_id: cveDetails.data.business_risk_id,
                id: cveName,
                justification: cveDetails.data.business_risk_justification
            }
        ]),
        kebabItemEditStatus(showStatusModal, [cveStatusDetails])
    ];
    const { rules } = data;
    const filterRuleValues = rules && rules.map(
        ({ rule_id: ruleId, description }) => ({ value: ruleId, label: description })
    );

    if (error?.hasError) {
        return (
            <React.Fragment>
                <Header title={cveName}/>
                <ErrorHandler code={error?.errorCode} />
            </React.Fragment>
        );
    } else {
        return (
            <React.Fragment>
                <CVEPageContext.Provider value={cveDetails && { isLoading: cveDetails.isLoading }}>
                    <Header
                        title={data.celebrity_name ? cveName + ' - ' + data.celebrity_name : cveName}
                        actions={kebabItems}
                        actionsOuiaId={'cve-actions'}
                        breadcrumbs={[
                            {
                                title: PATHS.cvesPage.title,
                                to: PATHS.cvesPage.to,
                                loaded: true
                            },
                            {
                                title: cveName,
                                isActive: true,
                                loaded: true
                            }
                        ]}
                        labels={[
                            <GroupedCVELabels
                                key="labels"
                                hasExploit={!!details.payload.data?.attributes.known_exploit}
                                hasRule={details.payload.data?.attributes.rules.length > 0}
                            />
                        ]}
                    >
                        <CVEDetailsPageSummary
                            changeExposedSystemsParameters={changeExposedSystemsParameters}
                            data={cveDetails}
                        />
                        <StatusModal />
                        <BusinessModal />
                    </Header>
                    <Main>
                        <SystemsExposedTable
                            cve={cveName}
                            filterRuleValues={filterRuleValues}
                            cveStatusDetails={cveStatusDetails}
                            methods={{ showStatusModal }}
                        />
                    </Main>
                </CVEPageContext.Provider>
            </React.Fragment>
        );
    }
};

CVEDetailsPage.propTypes = {
    match: propTypes.object
};

export default withRouter(CVEDetailsPage);
