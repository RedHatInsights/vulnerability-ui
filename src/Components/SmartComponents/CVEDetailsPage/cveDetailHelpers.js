import { getAffectedSystemsByCVE } from '../../../Helpers/APIHelper';

const getTotalItems = apiResult => (apiResult?.meta?.total_items || 0);

export const fetchHybridSystemsCounts = async (
    cveName,
    isEdgeParityEnabled,
    setHybridSystemsCounts
) => {
    try {
        const conventionalCount = await getAffectedSystemsByCVE({ id: cveName, host_type: 'rpmdnf', limit: 1 })
            .then(getTotalItems);

        const edgeCount = isEdgeParityEnabled
            ? await getAffectedSystemsByCVE({ id: cveName, host_type: 'edge', limit: 1 }).then(getTotalItems)
            : 0;

        setHybridSystemsCounts({
            conventionalCount,
            edgeCount,
            areCountsLoading: false
        });
    }
    catch {
        setHybridSystemsCounts({
            conventionalCount: 0,
            edgeCount: 0,
            areCountsLoading: false
        });
    }
};
