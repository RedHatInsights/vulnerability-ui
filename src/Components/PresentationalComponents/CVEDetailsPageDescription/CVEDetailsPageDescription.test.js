import React from 'react';
import CVEDetailsPageDescription from './CVEDetailsPageDescription';
import { render } from '@testing-library/react';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';
import { CVEPageContext } from '../../SmartComponents/CVEDetailsPage/CVEDetailsPage';

describe('CVEDetailsPageDescription component', () => {
    const cveAttributes = {
        business_risk: "Not defined",
        business_risk_id: 0,
        business_risk_justification: null,
        cvss3_metrics: "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:N/A:H",
        cvss3_score: "5.900",
        public_date: "12 May 2020",
        impact: "Important",
        status: "Resolved via mitigation",
        status_id: 6,
        synopsis: "CVE-2020-10711",
        description:
            'It was found that ghostscript could leak sensitive operators on the operand stack when a pseudo-operator' +
            ' pushes a subroutine. A specially crafted PostScript file could use this flaw to escape the -dSAFER' +
            ' protection in order to, for example, have access to the file system and execute commands.',
        systems_status_detail: { 0: 2 }
    };

    it('should render without props', () => {
        const { asFragment } = render(
            <IntlProvider locale="en">
                <CVEPageContext.Provider value={{ isLoading: false }}>
                    <CVEDetailsPageDescription cveAttributes={cveAttributes} />
                </CVEPageContext.Provider>
            </IntlProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
