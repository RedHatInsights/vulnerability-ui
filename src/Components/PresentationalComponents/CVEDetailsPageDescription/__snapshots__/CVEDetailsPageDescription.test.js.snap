// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVEDetailsPageDescription component should render without props 1`] = `
<injectIntl(CVEDetailsPageDescription)
  cveAttributes={
    Object {
      "business_risk": "Not defined",
      "business_risk_id": 0,
      "business_risk_justification": null,
      "cvss3_metrics": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:N/A:H",
      "cvss3_score": "5.900",
      "description": "It was found that ghostscript could leak sensitive operators on the operand stack when a pseudo-operator pushes a subroutine. A specially crafted PostScript file could use this flaw to escape the -dSAFER protection in order to, for example, have access to the file system and execute commands.",
      "impact": "Important",
      "public_date": "12 May 2020",
      "status": "Resolved via mitigation",
      "status_id": 6,
      "synopsis": "CVE-2020-10711",
      "systems_status_detail": Object {
        "0": 2,
      },
    }
  }
>
  <CVEDetailsPageDescription
    cveAttributes={
      Object {
        "business_risk": "Not defined",
        "business_risk_id": 0,
        "business_risk_justification": null,
        "cvss3_metrics": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:N/A:H",
        "cvss3_score": "5.900",
        "description": "It was found that ghostscript could leak sensitive operators on the operand stack when a pseudo-operator pushes a subroutine. A specially crafted PostScript file could use this flaw to escape the -dSAFER protection in order to, for example, have access to the file system and execute commands.",
        "impact": "Important",
        "public_date": "12 May 2020",
        "status": "Resolved via mitigation",
        "status_id": 6,
        "synopsis": "CVE-2020-10711",
        "systems_status_detail": Object {
          "0": 2,
        },
      }
    }
    intl={
      Object {
        "$t": [Function],
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "defaultRichTextElements": undefined,
        "fallbackOnEmptyString": true,
        "formatDate": [Function],
        "formatDateTimeRange": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatList": [Function],
        "formatListToParts": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "default.cancel": "Cancel",
          "default.delete": "Delete",
          "default.remove": "Remove",
          "default.save": "Save",
        },
        "onError": [Function],
        "onWarn": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
        "wrapRichTextChunksInFragment": undefined,
      }
    }
  >
    <Stack
      hasGutter={true}
    >
      <div
        className="pf-l-stack pf-m-gutter"
      >
        <StackItem
          className="pf-u-mt-sm"
        >
          <div
            className="pf-l-stack__item pf-u-mt-sm"
          >
            <Flex>
              <div
                className="pf-l-flex"
              >
                <FlexItem>
                  <div
                    className=""
                  >
                    Publish date
                    :
                  </div>
                </FlexItem>
                <FlexItem
                  grow={
                    Object {
                      "default": "grow",
                    }
                  }
                >
                  <div
                    className="pf-m-grow"
                  >
                    <WithLoader
                      isLoading={true}
                      size="sm"
                      variant="inlineSkeleton"
                    >
                      <Skeleton
                        isDark={false}
                        size="sm"
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <div
                          className="ins-c-skeleton ins-c-skeleton__sm"
                          style={
                            Object {
                              "display": "inline-block",
                            }
                          }
                        >
                           
                        </div>
                      </Skeleton>
                    </WithLoader>
                  </div>
                </FlexItem>
              </div>
            </Flex>
          </div>
        </StackItem>
        <StackItem>
          <div
            className="pf-l-stack__item"
          >
            <WithLoader
              isLoading={true}
              style={
                Object {
                  "height": "132px",
                  "width": "100%",
                }
              }
              variant="rectangle"
            >
              <Skeleton
                isDark={false}
                shape="square"
                size="md"
                style={
                  Object {
                    "height": "132px",
                    "width": "100%",
                  }
                }
              >
                <div
                  className="ins-c-skeleton ins-c-skeleton__md"
                  shape="square"
                  style={
                    Object {
                      "height": "132px",
                      "width": "100%",
                    }
                  }
                >
                   
                </div>
              </Skeleton>
            </WithLoader>
          </div>
        </StackItem>
        <StackItem
          className="pf-u-mt-sm pf-u-mb-md"
        >
          <div
            className="pf-l-stack__item pf-u-mt-sm pf-u-mb-md"
          />
        </StackItem>
      </div>
    </Stack>
  </CVEDetailsPageDescription>
</injectIntl(CVEDetailsPageDescription)>
`;
